go-controller/pkg/clustermanager/userdefinednetwork/controller_test.go

```diff
@@ -17,535 +19,854 @@ import (
 	"k8s.io/apimachinery/pkg/types"
 	informerfactory "k8s.io/client-go/informers"
 	corev1informer "k8s.io/client-go/informers/core/v1"
-	"k8s.io/client-go/kubernetes/fake"
 	"k8s.io/client-go/testing"
 	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	netv1 "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
+	netv1clientset "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/client/clientset/versioned"
 	netv1fakeclientset "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/client/clientset/versioned/fake"
 	netv1informerfactory "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/client/informers/externalversions"
 	netv1Informer "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/client/informers/externalversions/k8s.cni.cncf.io/v1"
 
 	udnv1 "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/userdefinednetwork/v1"
+	udnclient "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/userdefinednetwork/v1/apis/clientset/versioned"
 	udnfakeclient "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/userdefinednetwork/v1/apis/clientset/versioned/fake"
 	udninformerfactory "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/userdefinednetwork/v1/apis/informers/externalversions"
 	udninformer "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/userdefinednetwork/v1/apis/informers/externalversions/userdefinednetwork/v1"
 
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/clustermanager/userdefinednetwork/template"
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
 )
 
 var _ = Describe("User Defined Network Controller", func() {
 	var (
-		udnClient   *udnfakeclient.Clientset
-		nadClient   *netv1fakeclientset.Clientset
-		udnInformer udninformer.UserDefinedNetworkInformer
-		nadInformer netv1Informer.NetworkAttachmentDefinitionInformer
-		kubeClient  *fake.Clientset
-		podInformer corev1informer.PodInformer
+		cs                *util.OVNClusterManagerClientset
+		udnInformer       udninformer.UserDefinedNetworkInformer
+		nadInformer       netv1Informer.NetworkAttachmentDefinitionInformer
+		podInformer       corev1informer.PodInformer
+		namespaceInformer corev1informer.NamespaceInformer
```

Comment:
Using indexers from informers to make changes is a bit strange, you could instead use create/update/delete calls, e.g. for NAD you can do
`cs.NetworkAttchDefClient.K8sCniCncfIoV1().NetworkAttachmentDefinitions(namespace).Create(...)` same for all other objects, their clients are already present in the cs.

Reply:
Unfortunately, it doesnt seem to work well when tests relay on watch-factory, on NAD create/update/delete it seem the WF informers take some time to sync and have the actual new object state causing some tests to flake and other to fail.
I managed to work-around it by setting 50ms sleep right after NAD create/update/delete operation, but its nasty.
Using the WF informers directly in tests to simulate create/update/delete NAD does work though.

Reply:
I think all tests that manually call sync function have setup + sync + check result pattern. In this case, the easiest way to setup all informers with the right objects is to pass all udns/nads/etc to the `newTestController`, which will make sure all objects are propagated during watch factory start. Then, for example, test case "should fail when NAD owner-reference is malformed" may look like
```
                        udn := testUDN()
			mutatedNAD := testNAD()
			mutatedNAD.ObjectMeta.OwnerReferences = []metav1.OwnerReference{{Kind: "DifferentKind"}}

			c := newTestController(noopRenderNadStub(), udn, mutatedNAD)
			_, err := c.syncUserDefinedNetwork(udn)

			Expect(err).To(Equal(errors.New("foreign NetworkAttachmentDefinition with the desired name already exist [test/test]")))
```
which I thin makes it easier to read.
One more thing to make it work, you need to apply similar changes from https://github.com/ovn-org/ovn-kubernetes/commit/434b0590ce8c61ade75edc996b2f7f83d530f840  to the `fake_client.go`, to make sure NADs will be added with `GetOVNClientset`